sinks/gcm.go

```diff
@@ -55,112 +69,123 @@ func NewSink() (*GcmSink, error) {
-	for i := 0; i < len(metrics); i += step {
-		endIndex := i + step
-		if endIndex > len(metrics) {
-			endIndex = len(metrics)
-		}
-
-		err := self.driver.PushMetrics(metrics[i:endIndex])
+	for _, pod := range data.Pods {
+		err := self.podToMetrics(&pod)
 		if err != nil {
 			lastErr = err
 		}
 	}
+	err := self.rawContainersToMetrics(data.Containers)
+	if err != nil {
+		lastErr = err
+	}
+	err = self.rawContainersToMetrics(data.Machine)
+	if err != nil {
+		lastErr = err
+	}
 
 	return lastErr
 }
 
-// Concatenates a map of labels into a comma-separated key=value pairs.
-func labelsToString(labels map[string]string) string {
-	var buffer bytes.Buffer
-	first := true
-	for key, value := range labels {
-		if !first {
-			buffer.WriteString(",")
-		}
-		first = false
-		buffer.WriteString(key)
-		buffer.WriteString("=")
-		buffer.WriteString(value)
-	}
-	return buffer.String()
-}
-
-func (self *GcmSink) podToMetrics(pod *sources.Pod) []Metric {
-	metrics := make([]Metric, 0, len(pod.Containers))
-
+func (self *GcmSink) podToMetrics(pod *sources.Pod) error {
 	// Generate the labels.
 	labels := make(map[string]string)
 	labels[labelPodId] = pod.ID
-	labels[labelLabels] = labelsToString(pod.Labels)
+	labels[labelLabels] = gcm.LabelsToString(pod.Labels)
 	labels[labelHostname] = pod.Hostname
 
 	// Break the individual metrics from the container statistics.
+	var lastErr error
 	for _, container := range pod.Containers {
-		metrics = append(metrics, self.containerToMetrics(container, labels)...)
+		err := self.containerToMetrics(container, labels)
+		if err != nil {
+			lastErr = err
+		}
 	}
-	return metrics
-}
 
-func (self *GcmSink) rawContainersToMetrics(containers []sources.RawContainer) []Metric {
-	metrics := make([]Metric, 0, len(containers))
+	return lastErr
+}
 
+func (self *GcmSink) rawContainersToMetrics(containers []sources.RawContainer) error {
 	labels := make(map[string]string)
+	var lastErr error
 	for _, container := range containers {
 		labels[labelHostname] = container.Hostname
-		metrics = append(metrics, self.containerToMetrics(&container.Container, labels)...)
+		err := self.containerToMetrics(&container.Container, labels)
+		if err != nil {
+			lastErr = err
+		}
 	}
-	return metrics
+
+	return lastErr
 }
 
-func (self *GcmSink) containerToMetrics(container *sources.Container, labels map[string]string) []Metric {
+func (self *GcmSink) containerToMetrics(container *sources.Container, labels map[string]string) error {
 	labels[labelContainerName] = container.Name
 
 	// One metric value per data point.
-	metrics := make([]Metric, 0, len(container.Stats))
+	var lastErr error
+	labelsAsString := gcm.LabelsToString(labels)
 	for _, stat := range container.Stats {
 		// Add all supported metrics that have values.
+		m := make([]gcm.Metric, 0, len(self.supportedMetrics))
 		for _, supported := range self.supportedMetrics {
+			// GCM won't accept metrics older than the last point we pushed so ignore any stats
```

Comment:
We should figure out what is going on I just filed #107 to track that, but this makes this driver move forward. We can always remove it later.